### PR TITLE
Initial monitoring implementation

### PR DESCRIPTION
**Pull request by jim-minter**
Ideas:
* every cluster is randomly placed at create time into one of 256 buckets
* monitoring is horizontally scalable, active/active
* every monitor process advertises its liveness to the database
* one monitor holds a master lease at any one time; the master distributes ownership of the 256 buckets evenly across live monitor processes
* each monitor uses a Cosmos DB change feed to cache database state locally (like k8s list/watch)
* each monitor aims to check each cluster it owns every 5 minutes; it walks the local database state and distributes checking over lots of local goroutine workers
* monitoring stats are output to mdm via statsd

Back-of-envelope calculations:
* to support 50,000 clusters/RP with (say) 3 monitors, and check every cluster every 5 minutes, each monitor will need to retire 55 checks per second.
* if each check is allowed up to 30 seconds to run, that implies 1650 active goroutines per monitor.
* if each cluster's cached data model takes 2KB and each goroutine takes 2KB, memory usage per monitor would be around 103MB

### CODE REVIEW

**Comment by jim-minter** [comment]
Basic testing instructions:
```
$ . ./env
$ nc -U -l mdm_statsd.socket &
$ FEATURE_METRICS=true go run ./cmd/aro monitor
```

**Comment by jim-minter** [comment]
@pnasrat this passes a basic smoke test; ptal.

**Comment by pnasrat** [comment]
Just for context what other designs or existing open source solutions were considered, and what pros/cons for each approach were evaluated leading to this architecture?

File: pkg/database/monitors.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+	if doc.ID != strings.ToLower(doc.ID) {
+		return nil, fmt.Errorf("id %q is not lower case", doc.ID)
+	}
+
+	return c.c.Replace(ctx, doc.ID, doc, options)
+}
+
+func (c *monitors) Dequeue(ctx context.Context) (*api.MonitorDocument, error) {
+	docs, err := c.c.QueryAll(ctx, "", &cosmosdb.Query{
+		Query: `SELECT * FROM Monitors doc WHERE doc.id = "monitor" AND (doc.leaseExpires ?? 0) < GetCurrentTimestamp() / 1000`,
```
Comment (mjudeikis) [inline pkg/database/monitors.go]: `doc.id = "monitor"` is confusing. Can we document logic little bit?  Considering below we have `SELECT * FROM Monitors doc WHERE doc.id != "monitor"`

File: pkg/database/monitors.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+			continue
+		}
+		return doc, err
+	}
+
+	return nil, nil
+}
+
+func (c *monitors) ListBuckets(ctx context.Context) (buckets []int, err error) {
+	doc, err := c.get(ctx, "monitor")
```
Comment (mjudeikis) [inline pkg/database/monitors.go]: same as above. Confusing.

File: pkg/monitor/master.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+	"github.com/Azure/ARO-RP/pkg/api"
+)
+
+// master updates the monitor document with the list of buckets balanced between
+// registered monitors
+func (mon *monitor) master(ctx context.Context) error {
+	// if we know we're not the master, attempt to gain the lease on the monitor
+	// document
+	if !mon.isMaster {
+		doc, err := mon.db.Monitors.Dequeue(ctx)
```
Comment (mjudeikis) [inline pkg/monitor/master.go]: I think `docs.id = monitor` for lease master should be something `id=lease`

File: pkg/monitor/worker.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+// filler tracks the OpenShiftClusters change feed and keeps mon.docs
+// up-to-date.  We don't monitor clusters in Creating state, hence we don't add
+// them to mon.docs.  We also don't monitor clusters in Deleting state; when
+// this state is reached we delete from mon.docs
+func (mon *monitor) filler(ctx context.Context, log *logrus.Entry, stop <-chan struct{}) {
+	defer recover.Panic(log)
+
+	i := mon.db.OpenShiftClusters.ChangeFeed()
```
Comment (mjudeikis) [inline pkg/monitor/worker.go]: Does the changeFeed give all "previous" documents when it starts or just ones created/modified during the time we are listening?

Comment (jim-minter) [reply pkg/monitor/worker.go]: It is like list and watch in one go.  It brings you up-to-date from zero then gives you subsequent modifications.  I found  https://medium.com/@gary.strange/cosmosdb-change-feed-processing-c002448a6ca really useful on this.

File: pkg/database/monitors.go (hunk 1 above)
Comment (jim-minter) [reply pkg/database/monitors.go]: Yeah, the Dequeue name really doesn't help as well.  I'll have another go at the terminology and comments.

**Comment by jim-minter** [comment]
Trying to answer some of the above.  Initial goals include:
- to jump start our monitoring capabilities to get basic visibility quickly and enable rapid iteration
- to enable capabilities which cluster admins cannot significantly tamper with
- straightforward operational prerequisites (reliability, upgradeability, observability, basic scalability, state management, management in multiple regions, etc.)

To me, the first two imply external monitoring, but not to the exclusion of adding monitoring from *inside* the cluster as well as a complementary near-term goal.  Fundamentally I think both are required and fast.

Most of the tool touch points are pretty custom (custom data model in cosmos, custom networking, custom monitoring rules, Geneva customised statsd format) and I'm not aware of any open source tooling whose reuse would obviously make things better (this statement doesn't apply to the separate internal monitoring case).  Note: a lot of existing RP library and framework is reused by this initial implementation.

MSFT are looking at open sourcing the v3 monitoring core, but timescales are tight, and note that its touchpoints are different again (e.g. no CosmosDB).  If released its learnings can nonetheless be melded in later in principle.

Happy to discuss further; by call may be better than by GH issue though.

**Comment by mjudeikis** [comment]
I think this is fine. There is a little bit more complexities around change feed and how we construct buckets and distribute monitoring workload. I think that part could be simplified, but at a cost of that we would not have even distribution around workers.

**Comment by jim-minter** [comment]
clarifications applied and added docs/monitoring.md.  Will merge on green.

**Pull request merged**